### PR TITLE
Add error states

### LINKED ISSUE #19: Error Message / Retry option in the Recycler View Layout

### PR DESCRIPTION
**Pull request by roideuniverse**
closes #19

### CODE REVIEW

File: src/main/java/com/carrotcreative/recyclercore/widget/ProgressRecyclerViewLayout.java [hunk 1]
```diff
@@ -117,4 +124,131 @@ public void stopScroll()
...
+                mProgressBar.setVisibility(GONE);
+            }
+        }
+        else
+        {
+            restorePreviousViewState();
+        }
+    }
+
+    public void setErrorStateEnabled(View errorView, boolean show)
```
Comment (BrandonRomano) [inline src/main/java/com/carrotcreative/recyclercore/widget/ProgressRecyclerViewLayout.java]: I don't think I'm entirely sold on this interface.

I would almost rather there be a completely separate method called `setErrorStateView(View errorView)` which is only responsible for setting the error state view (Usually would be done in onCreate), and only have `setErrorStateEnabled(boolean show)` as the toggle method.

File: src/main/java/com/carrotcreative/recyclercore/widget/ProgressRecyclerViewLayout.java [hunk 2]
```diff
@@ -117,4 +124,131 @@ public void stopScroll()
...
+            mRecyclerView.setVisibility(mPrevViewVisibilityInstanceState.getRecyclerViewVisibility());
+            mProgressBar.setVisibility(mPrevViewVisibilityInstanceState.getProgressViewVisibility());
+            mErrorStateView.setVisibility(mPrevViewVisibilityInstanceState.getErrorViewVisibility());
+        }
+    }
+
+    /**
+     * Add Visibility annotation for lint checks.
+     */
+    @IntDef({VISIBLE, INVISIBLE, GONE})
```
Comment (BrandonRomano) [inline src/main/java/com/carrotcreative/recyclercore/widget/ProgressRecyclerViewLayout.java:223]: I've never seen this... What is it?

File: src/main/java/com/carrotcreative/recyclercore/widget/ProgressRecyclerViewLayout.java [hunk 3]
```diff
@@ -117,4 +124,131 @@ public void stopScroll()
...
+            restorePreviousViewState();
+        }
+    }
+
+    public void setErrorStateEnabled(View errorView, boolean show)
+    {
+        if(mErrorStateView == null)
+        {
+            mErrorStateView = errorView;
+            mContainer.addView(mErrorStateView);
```
Comment (BrandonRomano) [inline src/main/java/com/carrotcreative/recyclercore/widget/ProgressRecyclerViewLayout.java:185]: Ah, I really like the decision you made here to make it simply get added to the root FrameLayout.  That makes this really flexible.

I was thinking we were going to have a FrameLayout container inside of our old Relative Layout centered...

This is much better :100:

**Comment by BrandonRomano** [comment]
I have one comment about the interface, and also a question about `IntDef` (I'm pretty sure I've answered it myself from looking online, but would love to also hear it in your words).

This is looking really good though I like it!

File: src/main/java/com/carrotcreative/recyclercore/widget/ProgressRecyclerViewLayout.java (hunk 1 above)
Comment (roideuniverse) [inline src/main/java/com/carrotcreative/recyclercore/widget/ProgressRecyclerViewLayout.java]: :+1:  True, I agree with you.

File: src/main/java/com/carrotcreative/recyclercore/widget/ProgressRecyclerViewLayout.java (hunk 2 above)
Comment (roideuniverse) [inline src/main/java/com/carrotcreative/recyclercore/widget/ProgressRecyclerViewLayout.java:223]: So, I found this piece in the Android's View.java class, 

```
    @IntDef({VISIBLE, INVISIBLE, GONE})
    @Retention(RetentionPolicy.SOURCE)
    public @interface Visibility {}
```

```
@Visibility
    public int getVisibility() {
        return mViewFlags & VISIBILITY_MASK;
    }
```

They have added a check for the visibility of the view, so when we try to pass integer type 0, 8 for the visibility change, it works as a runtime but shows a warning during lint/compile time. 
So I added this to counter that. :smile:

Comment (BrandonRomano) [inline src/main/java/com/carrotcreative/recyclercore/widget/ProgressRecyclerViewLayout.java:223]: Gotcha, very cool!

Comment (roideuniverse) [inline src/main/java/com/carrotcreative/recyclercore/widget/ProgressRecyclerViewLayout.java:223]: Yeah, I really like it. Its like passing int as an argument at runtime but working like an enum during development.

**Comment by roideuniverse** [comment]
@BrandonRomano Addressed your comments. Check now.

File: src/main/java/com/carrotcreative/recyclercore/widget/ProgressRecyclerViewLayout.java [hunk 4]
```diff
@@ -117,4 +124,134 @@ public void stopScroll()
...
+    public boolean isErrorStateEnabled()
+    {
+        if(mErrorStateView != null && mErrorStateView.getVisibility() == View.VISIBLE)
+        {
+            return true;
+        }
+        return false;
+    }
+
+    public void setErrorStateEnabled(boolean enable)
```
Comment (BrandonRomano) [inline src/main/java/com/carrotcreative/recyclercore/widget/ProgressRecyclerViewLayout.java:152]: Could we add javadoc style comments to all of the public methods here?

I just noticed there's no indication that I have to call `setErrorView` before I call this.

Comment (roideuniverse) [inline src/main/java/com/carrotcreative/recyclercore/widget/ProgressRecyclerViewLayout.java:152]: You don't have to set the `setErrorView` before you call this method `isErrorStateEnabled`

**Comment by BrandonRomano** [comment]
Okay this is good :+1: !!

**Comment by roideuniverse** [comment]
Going to merge this in!

**Pull request merged**